### ISSUE TITLE
Radius - Message-Authenticator computation enhancement

### ISSUE THREAD

**Issue by Matheus-Garbelini** [secdev/scapy#1971]
#### Brief description

When creating a Access-Request Radius packet, the message-authenticator attribute `(RadiusAttr_Message_Authenticator)` is not handled.

As Radius `authenticator` is not calculated, but randomly generated by the requestor for Access-Request packets, a patch is proposed to correctly calculate the `RadiusAttr_Message_Authenticator` attribute.

As per RFC 2869, the calculation is done the same way as before, but the message-authenticator field must be zero filled.

#### Environment

-   Scapy version: `2.4.2`
-   Python version: `2.7.15+`
-   Operating System: `Ubuntu 16.04`

#### Proposed fix

```Diff
--- scapy/layers/radius.py	2019-04-03 18:02:16.913650460 +0800
+++ scapy/radius.py	2019-04-03 18:37:04.570306573 +0800
@@ -531,7 +531,7 @@
     val = 24
 
 
-def prepare_packed_data(radius_packet, packed_req_authenticator):
+def prepare_packed_data(radius_packet, packed_req_authenticator, clear_message_authenticator):
     """
     Pack RADIUS data prior computing the authentication MAC
     """
@@ -542,6 +542,8 @@
 
     packed_attrs = b''
     for attr in radius_packet.attributes:
+        if type(attr) == RadiusAttr_Message_Authenticator and clear_message_authenticator:
+            attr.value = bytearray(attr.len)
         packed_attrs += raw(attr)
 
     return packed_hdr + packed_req_authenticator + packed_attrs
@@ -1134,7 +1136,16 @@
         Computes the authenticator field (RFC 2865 - Section 3)
         """
 
-        data = prepare_packed_data(self, packed_request_auth)
+        data = prepare_packed_data(self, packed_request_auth, clear_message_authenticator=False)
+        radius_mac = hashlib.md5(data + shared_secret)
+        return radius_mac.digest()
+    
+    def compute_message_authenticator(self, packed_request, shared_secret):
+        """
+        Computes the authenticator field (RFC 2869 - Page 33)
+        """
+
+        data = prepare_packed_data(self, packed_request, clear_message_authenticator=True)
         radius_mac = hashlib.md5(data + shared_secret)
         return radius_mac.digest()

```

**Comment by gpotter2** [comment]
Hi ! Thanks for the report.
Would you mind submitting a PR directly ? The same applies for your other issue.

That would allow us to review it directly :smile: and check the tests.

Thanks !

**Comment by Matheus-Garbelini** [comment]
@gpotter2 Gonna do that by this weekend. By the way, scapy is working all the way up to WPA Enterprise now, so thanks again for the help on that KRACK issue.

**Comment by Matheus-Garbelini** [comment]
@gpotter2 see #1987

**Comment by guedou** [comment]
Fixed by #1987
